### PR TITLE
Parent children sync down target for iOS

### PR DESCRIPTION
**Pull request by wmathurin**
With associated tests

### CODE REVIEW

**Comment by huminzhi** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Code coverage under minimum of 80%</td>
    </tr>
  </tbody>
</table>

## Current coverage for SmartSync.framework is `74%`
Files changed | - | - 
--- | --- | ---
SFSyncDownTarget.m | `68%` | :warning:
SFSoqlSyncDownTarget.m | `90%` | :white_check_mark:
SFSyncTarget.m | `97%` | :white_check_mark:

---
## Current coverage for SalesforceAnalytics.framework is `41%`
✅ *No files affecting coverage found*

---
## Current coverage for SalesforceSDKCore.framework is `23%`
✅ *No files affecting coverage found*

---
## Current coverage for SmartStore.framework is `39%`
Files changed | - | - 
--- | --- | ---
SFSmartStore.m | `60%` | :warning:

---

> Powered by [xcov](https://github.com/nakiostudio/xcov)

<p align="right" data-meta="generated_by_CI-result">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: libs/SmartSync/SmartSync/Classes/Util/SFChildrenInfo.h [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+extern NSString * const kSFChildrenInfoSObjectTypePlural;
+extern NSString * const kSFChildrenInfoParentIdFieldName;
+
+/**
+ * Simple object to capture details of children in parent-child relationship
+ */
+@interface SFChildrenInfo : SFParentInfo
+
+@property (nonatomic, readonly) NSString* sobjectTypePlural;
```
Comment (khawkins) [inline libs/SmartSync/SmartSync/Classes/Util/SFChildrenInfo.h]: With net-new files, do we want to add nullability specifiers?  Interoperability with Swift seems like a first-class requirement these days.

File: libs/SmartSync/SmartSync/Classes/Util/SFParentChildrenSyncHelper.m [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+ DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
+ WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import "SFParentChildrenSyncHelper.h"
+#import <SmartStore/SFSmartStore.h>
+
+@interface SFSyncTarget ()
```
Comment (khawkins) [inline libs/SmartSync/SmartSync/Classes/Util/SFParentChildrenSyncHelper.m]: Why would we not define this in an internal header file of `SFSyncTarget`?  Redefining internal interfaces is usually done when you don't have control over the source you're consuming.

**Comment by khawkins** [comment]
Looks good over all.

File: libs/SmartSync/SmartSync/Classes/Util/SFChildrenInfo.h (hunk 1 above)
Comment (wmathurin) [inline libs/SmartSync/SmartSync/Classes/Util/SFChildrenInfo.h]: Good point. Will do.

File: libs/SmartSync/SmartSync/Classes/Util/SFParentChildrenSyncHelper.m (hunk 2 above)
Comment (wmathurin) [inline libs/SmartSync/SmartSync/Classes/Util/SFParentChildrenSyncHelper.m]: I was being lazy. Will do. (I wish Objective-C supported "protected" methods ;-).

**Pull request merged**